build-image.sh

```diff
@@ -49,14 +49,29 @@ ${BUILD_CMD} \
   --tag hadoop-testing/mysql:${PROJECT_VERSION} \
   "${SELF_DIR}/mysql" $@
 
+function build_ranger() {
+  if [ -f download/ranger-${RANGER_VERSION}-admin.tar.gz ]; then
+      echo "Skipping ranger build, using existing download/ranger-${RANGER_VERSION}-admin.tar.gz"
+      return
+  fi
+  mkdir -p build/tmp
+  tar -zxf download/apache-ranger-${RANGER_VERSION}.tar.gz -C build/tmp
+  cp build/ranger/* build/tmp/apache-ranger-${RANGER_VERSION}/
+  bash build/tmp/apache-ranger-${RANGER_VERSION}/build.sh
+  cp build/tmp/apache-ranger-${RANGER_VERSION}/target/ranger-${RANGER_VERSION}-admin.tar.gz download/ranger-${RANGER_VERSION}-admin.tar.gz
+}
+
 function build_hadoop_master_image() {
+  build_ranger
```

Comment:
it's not suitable to put such steps in the `build-image` phase.

my original thoughts are:

1. the common practice of building a Docker image is writing everything (including downloading artifacts) to Dockerfile, I finally found it's quite painful for developers especially those who have **limited** access to the Internet.

2. then I decided to separate those heavy repeatedly downloading steps, so that users are only required to download them one time, from any place, in any way.

3. users could also provide customized artifacts instead of using the official pre-built ones.

such a heavy building step breaks the above assumption, I would like to add a step called `build`. then the workflow will be

```
./download.sh
./build.sh
./build-image/sh
docker compose up
```